solver/src/settlement_submission/retry.rs

```diff
@@ -0,0 +1,136 @@
+    .filter_map(|gas_price_result| async move {
+        match gas_price_result {
+            Ok(gas_price) => {
+                tracing::debug!("estimated gas price {}", gas_price);
+                Some(gas_price)
+            }
+            Err(err) => {
+                tracing::error!("gas price estimation failed: {:?}", err);
+                None
+            }
+        }
+    });
+    transaction_retry::gas_price_increase::enforce_minimum_increase_and_cap(gas_price_cap, stream)
+}
+
+fn is_transaction_error(error: &ExecutionError) -> bool {
+    // This is the error as we've seen it on openethereum nodes. The code and error messages can
+    // be found in openethereum's source code in `rpc/src/v1/helpers/errors.rs`.
+    // TODO: check how this looks on geth and infura. Not recognizing the error is not a serious
+    // problem but it will make us sometimes log an error when there actually was no problem.
+    matches!(error, ExecutionError::Web3(Web3Error::Rpc(RpcError { code, .. })) if code.code() == -32010)
+}
+
+pub struct SettleResult(pub Result<(), MethodError>);
+impl TransactionResult for SettleResult {
+    fn was_mined(&self) -> bool {
+        if let Err(err) = &self.0 {
+            !is_transaction_error(&err.inner)
+        } else {
+            true
+        }
+    }
+}
+
+pub struct SettlementSender<'a> {
+    pub contract: &'a GPv2Settlement,
+    pub nonce: U256,
+    pub settlement: EncodedSettlement,
+}
+#[async_trait::async_trait]
+impl<'a> TransactionSending for SettlementSender<'a> {
+    type Output = SettleResult;
+    async fn send(&self, gas_price: f64) -> Self::Output {
+        tracing::info!("submitting solution transaction at gas price {}", gas_price);
+        let mut method = self
+            .contract
+            .settle(
+                self.settlement.tokens.clone(),
+                self.settlement.clearing_prices.clone(),
+                self.settlement.encoded_trades.clone(),
+                self.settlement.encoded_interactions.clone(),
+                Vec::new(),
+            )
+            .nonce(self.nonce)
+            .gas(MAX_GAS.into())
+            .gas_price(GasPrice::Value(U256::from_f64_lossy(gas_price)));
+        method.tx.resolve = Some(ResolveCondition::Confirmed(ConfirmParams::mined()));
+        let result = method.send().await.map(|_| ());
+        SettleResult(result)
+    }
+}
+
+// We never send cancellations but we still need to have types that implement the traits.
+pub struct CancelResult;
+impl TransactionResult for CancelResult {
+    fn was_mined(&self) -> bool {
+        unreachable!()
+    }
+}
+
+pub struct CancelSender;
+#[async_trait::async_trait]
+impl TransactionSending for CancelSender {
+    type Output = CancelResult;
+    async fn send(&self, _gas_price: f64) -> Self::Output {
+        unreachable!()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use jsonrpc_core::ErrorCode;
+
+    #[test]
+    fn test_submission_result_was_mined() {
```

Comment:
Does this test somehow include a retry? If so, I am unable to tell how.

Reply:
it does not and is not supposed to. just tests that detection of was_mined works.